src/coprocessor/cache.rs

```diff
@@ -0,0 +1,493 @@
+                        // Region's version is not same as cached evict it
+                        if rentry.version != entry.version {
+                            Some(())
+                        } else {
+                            None
+                        }
+                    }
+                }
+            }
+        };
+        if opt.is_some() {
+            self.remove(k);
+        }
+    }
+
+    pub fn get_region_version_and_cache_entry(
+        &mut self,
+        region_id: u64,
+        k: &str,
+    ) -> (u64, Option<&Vec<u8>>) {
+        (self.get_region_version(region_id), self.get(region_id, k))
+    }
+
+    pub fn get_region_version(&self, region_id: u64) -> u64 {
+        match self.regions.get(&region_id) {
+            None => 0,
+            Some(item) => item.version,
+        }
+    }
+
+    pub fn get(&mut self, region_id: u64, k: &str) -> Option<&Vec<u8>> {
+        if !self.is_region_cache_enabled(region_id) {
+            return None;
+        }
+
+        self.check_evict_key(region_id, k);
+        if let Some(entry) = self.map.get_refresh(k) {
+            Some(&entry.result)
+        } else {
+            None
+        }
+    }
+
+    pub fn remove(&mut self, k: &str) {
+        let regions = &mut self.regions;
+        let option = self.map.remove(k);
+        match option {
+            None => (),
+            Some(entry) => {
+                let region_id: u64 = entry.region_id;
+                let opt = match regions.get_mut(&region_id) {
+                    None => None,
+                    Some(node) => {
+                        // Delete from region cache entry list
+                        node.cached_items.remove(k);
+                        if !node.cached_items.is_empty() && node.version != 1 {
+                            Some(())
+                        } else {
+                            None
+                        }
+                    }
+                };
+                if opt.is_some() {
+                    regions.remove(&region_id);
+                };
+                self.size -= entry.size;
+            }
+        };
+    }
+
+    pub fn evict_region(&mut self, region_id: u64) {
+        let keys = match self.regions.get_mut(&region_id) {
+            None => None,
+            Some(region) => {
+                region.version += 1;
+                let mut keys: Vec<DistSQLCacheKey> = Vec::new();
+                for (key, _) in (&region.cached_items).iter() {
+                    keys.push(key.to_string());
+                }
+                Some(keys)
+            }
+        };
+        match keys {
+            None => {
+                let entry = RegionDistSQLCacheEntry {
+                    version: 1,
+                    enable: true,
+                    cached_items: HashMap::new(),
+                };
+                self.regions.insert(region_id, entry);
+            }
+            Some(keys) => for i in keys {
```

Comment:
It seems too complicated for the reviewers, why not process new region and old region more  independently like:
```
if new region {
   init..
   return
}
then..
```

Reply:
I try to change to this form but got `second mutable borrow` error.

Reply:
May this doc help https://doc.rust-lang.org/nomicon/lifetimes.html @blacktear23 And try to make the code more friendly for the reviewers?